### PR TITLE
feat: add filter to swap

### PR DESCRIPTION
**Pull request by sarneijim**
### 📝 Description

Add filters to swap

### ❓ Context

- **Impacted projects**: 
  - LLD
- **Linked resource(s)**: 
  -  [swap] As a user on desktop, I want to be able to filter Centralised vs Decentralised providers https://ledgerhq.atlassian.net/browse/LIVE-3730
  - [swap] As a user on desktop, I want to be able to filter floating vs fixed rates quotes https://ledgerhq.atlassian.net/browse/LIVE-3785

### ✅ Checklist

- [ ] **Test coverage** 
- [ ] **Atomic delivery** 
- [ ] **No breaking changes** 

### 📸 Demo

Screenshot 2022-09-26 at 17 47 14

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

**Review by gre — COMMENTED** [review]
is there a way to test these new screens states? with mocks?

**Comment by sarneijim** [comment]
> is there a way to test these new screens states? with mocks?

These test are skipped. @ggilchrist-ledger has another jira to update these tests.

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  &:active,
+  &:hover {
+    color: ${(p) => (p.selected ? p.theme.colors.primary.c80 : p.theme.colors.neutral.c100)};
+  }
+`;
+
+export default function ProviderRate({ onClick }: Props) {
+  const [filter, setFilter] = useState([]);
+
+  const updateFilter = (type) => {
```
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx:44]: use a `useCallback` will avoid the `<Btn>` to always rerender (as the updateFilter prop will always change ref)

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    } else {
+      setFilter([type, ...filter]);
+      newFilter = [type, ...filter];
+    }
+    if ([FILTER.centralised, FILTER.decentralised].every((elem) => newFilter.includes(elem))) {
+      newFilter = newFilter.filter((f) => ![FILTER.centralised, FILTER.decentralised].includes(f));
+    }
+    if ([FILTER.fixed, FILTER.float].every((elem) => newFilter.includes(elem))) {
+      newFilter = newFilter.filter((f) => ![FILTER.fixed, FILTER.float].includes(f));
+    }
```
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx:58]: i'm a bit confused why you need both the "newFilter" internal let variable which diverges from the `filter` **state**. 

the code also isn't so easy to understand so maybe worth having more documentation (or maybe there is a simpler way to achieve what's done?)

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/index.jsx [hunk 3]
```diff
@@ -70,13 +76,31 @@ export default function ProviderRate({
...
     [updateSelection],
   );
 
+  useEffect(() => {
+    if (filter.includes(FILTER.decentralised)) {
+      setDexRate(DEX_PROVIDERS[0]);
+    } else {
+      if (filter.includes(FILTER.float)) {
+        const selectedRate = rates.find((rate) => rate.tradeMethod === FILTER.float);
+        setRate(selectedRate || rates[0]);
```
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/index.jsx:92]: dumb question but is it possible that rates[0] is undefined? is setRate ready to receive an undefined value?

**Review by gre — COMMENTED** [review]
is this covered by tests?
some code feedback

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx (hunk 1 above)
Comment (sarneijim) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx]: Done!

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx (hunk 2 above)
Comment (sarneijim) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/Filter.jsx]: I have had some comments to clarify.

If centralised & decentralised filters are active, these 2 filters will not be applied
If fixed & float filters are active, these 2 filters will not be applied

(for that reason the status of the component not represent exactly the filter itself)

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/index.jsx (hunk 3 above)
Comment (sarneijim) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/index.jsx]: Good catch! 
I have add a empty object if the value is undefined and I have unify the empty status (I had missing delete a empty state legacy)

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/index.jsx [hunk 4]
```diff
@@ -70,13 +76,29 @@ export default function ProviderRate({
...
+      let selectedRate;
+      if (filter.includes(FILTER.float)) {
+        selectedRate = rates.find((rate) => rate.tradeMethod === FILTER.float);
+      } else if (filter.includes(FILTER.fixed)) {
+        selectedRate = rates.find((rate) => rate.tradeMethod === FILTER.fixed);
+      }
+      setRate(selectedRate || (rates && rates[0]) || {});
+    }
+    setEmptyState(!providerRef.current?.children.length);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/index.jsx:99]: any chance rates is changing over time? I imagine you only care about the "initial rates" then? as you need to actively ignore the dep?

**Review by gre — APPROVED** [review]

**Comment by ggilchrist-ledger** [comment]
The 'no results found' doesn't look right @sarneijim 
image

**Comment by ggilchrist-ledger** [comment]
This should be 'Floating' rate @sarneijim 
image

**Comment by ggilchrist-ledger** [comment]
Also I found a funky bug @sarneijim:
1. Generate a quote
2. Remove the quote
3. Exchange is still clickable for a number of seconds (it eventually becomes unclickable)

You can then cause a crash from this point if you continue with the nano


https://user-images.githubusercontent.com/87659842/193594542-067b462f-d4dd-4992-aedf-a74d6a6567eb.mov

**Comment by ggilchrist-ledger** [comment]
LGTM! @sarneijim

**Pull request merged**